feat: Import a page from zip file

Mukund-Tandon opened the pull request:
<!---
Thank you for submitting a pull request to AppFlowy. The team will dedicate their best efforts to reviewing and approving your pull request. If you have any questions about the project or feedback for us, please join our [Discord](https://discord.gg/wdjWUXXhtw).
-->

<!---
If your pull request adds a new feature, please drag and drop a video into this section to showcase what you've done! If not, you may delete this section.
-->

### Feature Preview
This PR is WIP
Current progress - 

https://github.com/AppFlowy-IO/AppFlowy/assets/71614009/e506ec30-7299-4807-a2cf-cb7c445cd253


<!---
List at least one issue here that this PR addresses. If it fixes the issue, please use the [fixes](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests) keyword to close the issue. For example:
fixes https://github.com/AppFlowy-IO/AppFlowy/pull/2106
-->
Fixes #2911 
---

<!---
Before you mark this PR ready for review, run through this checklist!
-->

#### PR Checklist

- [ ] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [X] I've listed at least one issue that this PR fixes in the description above.
- [ ] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [ ] All existing tests are passing.

Mukund-Tandon commented:
@LucasXu0  Can you please review this PR. 
Many file are changed because I ran dart format .
main changes are in frontend/appflowy_flutter/lib/workspace/presentation/home/menu/app/header/import/import_panel.dart
and frontend/appflowy_flutter/lib/workspace/presentation/home/menu/app/header/import/import_type.dart files

LucasXu0 commented:
@Mukund-Tandon Please add some tests to cover this feature. And the CI will re-run automatically after you push some new codes/tests.

Mukund-Tandon commented:
> @Mukund-Tandon Please add some tests to cover this feature. And the CI will re-run automatically after you push some new codes/tests.

Ok 👍

Mukund-Tandon commented:
@LucasXu0 ,I while writing unit test for this PR I am stuck on some error 
The test code - 
```dart
import 'dart:typed_data';
import 'package:appflowy/startup/deps_resolver.dart';
import 'package:appflowy/startup/startup.dart';
import 'package:appflowy/workspace/application/settings/application_data_storage.dart';
import 'package:appflowy/workspace/presentation/home/menu/app/header/import/import_panel.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:get_it/get_it.dart';
import 'package:archive/archive.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();
  final getIt = GetIt.instance;

  setUpAll(() async {
    DependencyResolver.resolve(getIt, IntegrationMode.unitTest);
    MockApplicationDataStorage.initialPath = 'path/to/image';
  });
  test('extractImagePath should return the correct image path', () {
    const inputText = "![image](path/to/image.png)";
    final result = extractImagePath(inputText);
    const expectedImagePath = "path/to/image.png";
    expect(result, expectedImagePath);
  });

  test('extractImagePath should return null if no image path is found', () {
    const inputText = "This is an example text without an image";
    final result = extractImagePath(inputText);
    expect(result, isNull);
  });
  test('saveFileLocally saves the file and returns its path', () async {
    final images = [
      ArchiveFile('assetName.png', 3, Uint8List.fromList([0, 1, 2]))
    ];
    const assetName = 'assetName.png';

    final path = await saveFileLocally(images, assetName);

    expect(path!.startsWith('path/to/image/images'), isTrue);
    expect(path.endsWith('.png'), isTrue);
  });
} 
```

Is this the write way to write test saveFileLocally ?. I am getting this  error on the test -"saveFileLocally saves the file and returns its path"
<img width="699" alt="Screenshot 2023-07-16 at 7 37 32 AM" src="https://github.com/AppFlowy-IO/AppFlowy/assets/71614009/c67ee29b-97f9-406d-8f95-b8b0a9428974">

LucasXu0 commented:
> @LucasXu0 , while writing unit test for this PR I am stuck on some error The test code -
> 
> ```dart
> import 'dart:typed_data';
> import 'package:appflowy/startup/deps_resolver.dart';
> import 'package:appflowy/startup/startup.dart';
> import 'package:appflowy/workspace/application/settings/application_data_storage.dart';
> import 'package:appflowy/workspace/presentation/home/menu/app/header/import/import_panel.dart';
> import 'package:flutter/widgets.dart';
> import 'package:flutter_test/flutter_test.dart';
> import 'package:get_it/get_it.dart';
> import 'package:archive/archive.dart';
> 
> void main() {
>   TestWidgetsFlutterBinding.ensureInitialized();
>   final getIt = GetIt.instance;
> 
>   setUpAll(() async {
>     DependencyResolver.resolve(getIt, IntegrationMode.unitTest);
>     MockApplicationDataStorage.initialPath = 'path/to/image';
>   });
>   test('extractImagePath should return the correct image path', () {
>     const inputText = "![image](path/to/image.png)";
>     final result = extractImagePath(inputText);
>     const expectedImagePath = "path/to/image.png";
>     expect(result, expectedImagePath);
>   });
> 
>   test('extractImagePath should return null if no image path is found', () {
>     const inputText = "This is an example text without an image";
>     final result = extractImagePath(inputText);
>     expect(result, isNull);
>   });
>   test('saveFileLocally saves the file and returns its path', () async {
>     final images = [
>       ArchiveFile('assetName.png', 3, Uint8List.fromList([0, 1, 2]))
>     ];
>     const assetName = 'assetName.png';
> 
>     final path = await saveFileLocally(images, assetName);
> 
>     expect(path!.startsWith('path/to/image/images'), isTrue);
>     expect(path.endsWith('.png'), isTrue);
>   });
> } 
> ```
> 
> Is this the right way to write test for saveFileLocally ?. I am getting this error on the test -"saveFileLocally saves the file and returns its path" <img alt="Screenshot 2023-07-16 at 7 37 32 AM" width="699" src="https://user-images.githubusercontent.com/71614009/253781479-c67ee29b-97f9-406d-8f95-b8b0a9428974.png">

Please write an integration test for it and you can refer to the `import_files_test.dart file`.

Here're the steps.

1. mock clicking the import button.
2. mock the file picker service to return the resources that added in `assets/test/...`
3. mock clicking the import page button
4. check the result.

Mukund-Tandon commented:
> > @LucasXu0 , while writing unit test for this PR I am stuck on some error The test code -
> > ```dart
> > import 'dart:typed_data';
> > import 'package:appflowy/startup/deps_resolver.dart';
> > import 'package:appflowy/startup/startup.dart';
> > import 'package:appflowy/workspace/application/settings/application_data_storage.dart';
> > import 'package:appflowy/workspace/presentation/home/menu/app/header/import/import_panel.dart';
> > import 'package:flutter/widgets.dart';
> > import 'package:flutter_test/flutter_test.dart';
> > import 'package:get_it/get_it.dart';
> > import 'package:archive/archive.dart';
> > 
> > void main() {
> >   TestWidgetsFlutterBinding.ensureInitialized();
> >   final getIt = GetIt.instance;
> > 
> >   setUpAll(() async {
> >     DependencyResolver.resolve(getIt, IntegrationMode.unitTest);
> >     MockApplicationDataStorage.initialPath = 'path/to/image';
> >   });
> >   test('extractImagePath should return the correct image path', () {
> >     const inputText = "![image](path/to/image.png)";
> >     final result = extractImagePath(inputText);
> >     const expectedImagePath = "path/to/image.png";
> >     expect(result, expectedImagePath);
> >   });
> > 
> >   test('extractImagePath should return null if no image path is found', () {
> >     const inputText = "This is an example text without an image";
> >     final result = extractImagePath(inputText);
> >     expect(result, isNull);
> >   });
> >   test('saveFileLocally saves the file and returns its path', () async {
> >     final images = [
> >       ArchiveFile('assetName.png', 3, Uint8List.fromList([0, 1, 2]))
> >     ];
> >     const assetName = 'assetName.png';
> > 
> >     final path = await saveFileLocally(images, assetName);
> > 
> >     expect(path!.startsWith('path/to/image/images'), isTrue);
> >     expect(path.endsWith('.png'), isTrue);
> >   });
> > } 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is this the right way to write test for saveFileLocally ?. I am getting this error on the test -"saveFileLocally saves the file and returns its path" <img alt="Screenshot 2023-07-16 at 7 37 32 AM" width="699" src="https://user-images.githubusercontent.com/71614009/253781479-c67ee29b-97f9-406d-8f95-b8b0a9428974.png">
> 
> Please write an integration test for it and you can refer to the `import_files_test.dart file`.
> 
> Here're the steps.
> 
> 1. mock clicking the import button.
> 2. mock the file picker service to return the resources that added in `assets/test/...`
> 3. mock clicking the import page button
> 4. check the result.

Ok 👍,Thanks

Mukund-Tandon commented:
@LucasXu0  I have added the integration test please can you review it. 

Integration Test 

https://github.com/AppFlowy-IO/AppFlowy/assets/71614009/08d1ea7c-13e8-4c27-b204-b40cf17f34f5

Feature preview-

https://github.com/AppFlowy-IO/AppFlowy/assets/71614009/57ded275-5fca-499e-9a72-39a536131923

LucasXu0 commented:
> @LucasXu0, I noticed that AppFlowy supports inline references to pages. Can I also add subpages while importing a page from Notion in this pull request?

I suggest that you submit a new PR to support that, as it depends on a new API being developed by @AmanNegi.

And please rebase the main branch.

Mukund-Tandon commented:
> > @LucasXu0, I noticed that AppFlowy supports inline references to pages. Can I also add subpages while importing a page from Notion in this pull request?
> 
> I suggest that you submit a new PR to support that, as it depends on a new API being developed by @AmanNegi.
> 
> And please rebase the main branch.

Ok, so there is nothing more to work on this PR?

Mukund-Tandon commented:
@LucasXu0 , I have integrated latest appflowy-editor and this is how it is working -

https://github.com/AppFlowy-IO/AppFlowy/assets/71614009/449ddca8-7bf5-42c6-8cc8-9ea1e085faf1


But I was improving the integration test by detecting if there is image imprted from markdown file but the test is failing and it looks like I am unable to open page- "AppFlowy Test"-
The test code - 

```dart
import 'dart:io';

import 'package:appflowy_editor/appflowy_editor.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';

import 'util/mock/mock_file_picker.dart';
import 'util/util.dart';
import 'package:path/path.dart' as p;

void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();
  group('import files', () {
    testWidgets('import page from notion', (tester) async {

      const pageName = 'AppFlowy Test';
      final context = await tester.initializeAppFlowy();
      await tester.tapGoButton();

      // expect to see a readme page
      tester.expectToSeePageName(readme);

      await tester.tapAddButton();
      await tester.tapImportButton();

      
      final paths = <String>[];
      final ByteData data = await rootBundle
          .load('assets/test/workspaces/import_page_from_notion_test.zip');
      final path = p.join(
        context.applicationDataDirectory,
        'import_page_from_notion_test.zip',
      );
      paths.add(path);
      final file = File(path);
      await file.writeAsBytes(
        data.buffer.asUint8List(data.offsetInBytes, data.lengthInBytes),
      );
      // mock get files

      expect(find.widgetWithText(Card, 'Page'), findsNothing);
      await tester.tapButtonWithName('Import from Notion');
      expect(find.widgetWithText(Card, 'Page'), findsOneWidget);
      await tester.tapButtonWithName('Page');
      expect(find.text('Import Notion Page'), findsOneWidget);
      await mockPickFilePaths(
        paths: paths,
      );
      await tester.tapButtonWithName('Upload zip file');
      tester.expectToSeePageName(pageName);
      await tester.openPage(pageName);
      expect(
        tester.editor.getCurrentEditorState().getNodeAtPath([4])!.type,
        ImageBlockKeys.type,
      );
    });
  });
}
```

Test - 

https://github.com/AppFlowy-IO/AppFlowy/assets/71614009/6ca634a2-daea-4976-ac2e-d2e5e27bea81

LucasXu0 commented:
@Mukund-Tandon Okay. I'm checking out your branch and will fix the test issues.

LucasXu0 commented:
- moved the import from notion panel into a widget class `import_from_notion_widget.dart`
-

Mukund-Tandon commented:
@LucasXu0 , I figured out why the test was not working . 
It was becuase when we were call `await tester.openPage(pageName);` this statement the import panel was open so this statement only close the import panel . To make the test run I called this   `await tester.openPage(pageName);`again and it opened the page and all the test passed -
Test - 

https://github.com/AppFlowy-IO/AppFlowy/assets/71614009/b3fba866-6c04-47ee-826c-422ee52949e3

Code -
```dart
import 'dart:io';

import 'package:appflowy_editor/appflowy_editor.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';

import 'util/mock/mock_file_picker.dart';
import 'util/util.dart';
import 'package:path/path.dart' as p;

void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();
  group('import files', () {
    testWidgets('import page from notion', (tester) async {

      const pageName = 'AppFlowy Test';
      final context = await tester.initializeAppFlowy();
      await tester.tapGoButton();

      // expect to see a readme page
      tester.expectToSeePageName(readme);

      await tester.tapAddButton();
      await tester.tapImportButton();

      
      final paths = <String>[];
      final ByteData data = await rootBundle
          .load('assets/test/workspaces/import_page_from_notion_test.zip');
      final path = p.join(
        context.applicationDataDirectory,
        'import_page_from_notion_test.zip',
      );
      paths.add(path);
      final file = File(path);
      await file.writeAsBytes(
        data.buffer.asUint8List(data.offsetInBytes, data.lengthInBytes),
      );
      // mock get files

      expect(find.widgetWithText(Card, 'Page'), findsNothing);
      await tester.tapButtonWithName('Import from Notion');
      expect(find.widgetWithText(Card, 'Page'), findsOneWidget);
      await tester.tapButtonWithName('Page');
      expect(find.text('Import Notion Page'), findsOneWidget);
      await mockPickFilePaths(
        paths: paths,
      );
      await tester.tapButtonWithName('Upload zip file');
      tester.expectToSeePageName(pageName);
      await tester.openPage(pageName);
      //the above one openPage command closes the import panel
      await tester.openPage(pageName);
      expect(
        tester.editor.getCurrentEditorState().getNodeAtPath([4])!.type,
        ImageBlockKeys.type,
      );
    });
  });
}
```

Mukund-Tandon commented:
I have also changed the import_page_from_notion_test.zip file

LucasXu0 commented:
> I have also changed the import_page_from_notion_test.zip file

This markdown inside the zip is too simple. Please add more formatted text, such as headings, bulleted lists, numbered lists, and so on, and check the page to ensure they are imported successfully.

Mukund-Tandon commented:
> > I have also changed the import_page_from_notion_test.zip file
> 
> This markdown inside the zip is too simple. Please add more formatted text, such as headings, bulleted lists, numbered lists, and so on, and check the page to ensure they are imported successfully.

Ok 👍

LucasXu0 commented:
> I have also changed the import_page_from_notion_test.zip file

By the way, you must add the test to the runner.dart, otherwise it won't run. https://github.com/AppFlowy-IO/AppFlowy/pull/2981/commits/3fcad4a6a2e754dc3d986887628199cfbda636d5

Mukund-Tandon commented on frontend/appflowy_flutter/lib/workspace/presentation/home/menu/app/header/import/importer/notion_importer.dart:
```diff
@@ -0,0 +1,136 @@
+    final lines = markdown.split('\n');
+    final result = <String>[];
+    for (final line in lines) {
+      if (line.isEmpty) {
```
@LucasXu0 , This if condition will remove empty lines. If the original Notion page contains empty spaces between two paragraphs, they will be removed. Is this an intentional change, or can I remove this?

LucasXu0 replied:
It's an intentional change. I imported a markdown from Notion and found that there are so many blank lines. So, I added this condition. By the way, I updated the translations, but I forgot to update the test. Please help me update the test.

Mukund-Tandon replied:
Ok . I will update the test zip file with more contents and also update the integration test.

Mukund-Tandon commented:
@LucasXu0  , while writing the test ,I found out appflowy-editor do not support importing numbered list. It is treating it as a simple paragraph.
I can raise a PR to solve this

Mukund-Tandon commented:
Also there was an Overflow error occuring, which I have fixed -
Before -
 
<img width="1374" alt="Screenshot 2023-07-20 at 2 01 39 PM" src="https://github.com/AppFlowy-IO/AppFlowy/assets/71614009/108d4347-e021-4b2d-940b-faf4de0a746d">
After

<img width="1374" alt="Screenshot 2023-07-20 at 2 06 19 PM" src="https://github.com/AppFlowy-IO/AppFlowy/assets/71614009/eea94ddf-90bc-4029-ad97-572f42583f38">

Mukund-Tandon commented:
@LucasXu0 , I have pushed some changes in my branch ,commits  are being reflected on my github  https://github.com/Mukund-Tandon/AppFlowy/tree/import_page_from_zip_file_2911. but not here. 
Should I close this PR and open a new one?

Mukund-Tandon commented:
@LucasXu0 , I have updated the appflowy-editor git ref to use the latest and have also added numbered list as part of the integration test for importing notion page

Mukund-Tandon commented:
@LucasXu0 , is there anything more that needs to be done on this PR?

Mukund-Tandon commented:
@LucasXu0 , is there anything else that needs to be done on this PR?

Mukund-Tandon commented:
After resolving the merge conflicts the integration test for importing notion page was failing. 
When we import a notion page it doesn't open the page(if the page is closed) unlike when we import a simple markdown. So I have fixed that and now when we import the notion page the page in which we import gets open(if it is closed) . This was also the reason for the failing integration test, since the page was not getting open and the test was expecting to see the imported page name on screen ,but could not see it.

https://github.com/AppFlowy-IO/AppFlowy/assets/71614009/889afec5-592d-4f25-a624-7dfef8d554b2

Mukund-Tandon commented:
I have added support for importing subpages of the page being imported -

https://github.com/AppFlowy-IO/AppFlowy/assets/71614009/c38b06c4-d4b8-45d9-8a2c-253af7fafaa8

Mukund-Tandon commented:
I have updated the integration test for importing the notion page -

https://github.com/AppFlowy-IO/AppFlowy/assets/71614009/1021dcff-8975-4951-bc3a-a0d5097f814b
